### PR TITLE
Poi implementation

### PR DESCRIPTION
**Pull request by Oggal**
Builds Initial Implementation of Terrain Noise Modifiers and POI Objects

POI are represented as scriptable objects, a new component  tracks these and provides them to WoldGen at time of generation

### CODE REVIEW

File: Runtime/Scripts/ProcPrefabPOI.cs
Comment (Oggal) [inline Runtime/Scripts/ProcPrefabPOI.cs:1]: This File should be removed. The system has been redesigned and no longer uses this file

File: Runtime/Scripts/TerrainNoiseModifier.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        float distance = Vector2.Distance(new Vector2(Position.x, Position.z), new Vector2(pointX, pointY));
+        if (distance > outerRadius + innerRadius)
+        {
+            return 0;
+        }
+        else if(distance < innerRadius)
+        {
+            return 1;
+        }
+        return 1 - (distance - innerRadius) / (outerRadius);
```
Comment (Oggal) [inline Runtime/Scripts/TerrainNoiseModifier.cs]: Could this be simplified into a clamp?

File: Runtime/Scripts/TerrainNoiseModifier.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    public virtual Vector3 GetPosition()
+    {
+        return Position;
+    }
+
+    public virtual void Pregeneration() { }
+
+    public UnityEvent OnSpawned;
+    public UnityEvent OnDespawned;
```
Comment (Oggal) [inline Runtime/Scripts/TerrainNoiseModifier.cs]: Are these events referenced anywhere?

File: Samples/POI_Examples/Placeholder_POI 3.asset
Comment (Oggal) [inline Samples/POI_Examples/Placeholder_POI 3.asset:1]: Check these assests, Ensure they're based on the right script.

File: Runtime/Components/WorldGen.cs [hunk 3]
```diff
@@ -126,65 +127,68 @@ public void ClearChildren()
...
 
-		//Actually Build the Tiles
-		for (int y = Radius * -1; y <= Radius; y++)
+        //Actually Build the Tiles
+        for (int y = Radius * -1; y <= Radius; y++)
         {
             for (int x = Radius * -1; x <= Radius; x++)
             {
-				int id = (Radius - y) * TRadius + (x + Radius);
+                int id = (Radius - y) * TRadius + (x + Radius);
```
Comment (Oggal) [inline Runtime/Components/WorldGen.cs:153]: So looking at this line I see an imediate problem, need to rename either Radius or TRadius, What is TRadius and how does it differ?

File: Runtime/Components/WorldGen.cs [hunk 4]
```diff
@@ -198,14 +202,14 @@ private void BuildMesh(int Tx, int Ty, GameObject HoldsMesh)
 		}
 	}
 
-
-    private float getRatio(float pX,float pY,out float height)
+    private float getRatio(float pX, float pY, out float height)
     {
 		height = 0;
-		if (UseCityGrid)
+		if (SpawnPOIs && GetComponent<POI_Gen>() != null)
```
Comment (Oggal) [inline Runtime/Components/WorldGen.cs:206]: Do we want to keep SpawnPOIs, or rely on presence of POI_Gen? Don't feel like we should do both...

File: Runtime/Components/WorldGen.cs [hunk 5]
```diff
@@ -198,14 +202,14 @@ private void BuildMesh(int Tx, int Ty, GameObject HoldsMesh)
...
 		{
+			TerrainNoiseModifier[] Mods =GetComponent<POI_Gen>()?.Mods;
 			float ratio = 0;
-			foreach(TerrainNoiseModifier TNM in Modifiers)
+			foreach(TerrainNoiseModifier TNM in Mods)
```
Comment (Oggal) [inline Runtime/Components/WorldGen.cs]: The closeness of Mods and Modifies has already caused issues, are both still present in WorldGen?

File: Runtime/Components/WorldGen.cs [hunk 6]
```diff
@@ -221,10 +225,11 @@ private float getRatio(float pX,float pY,out float height)
 	private float getRatio(float pX, float pY)
```
Comment (Oggal) [inline Runtime/Components/WorldGen.cs:223]: Repeated Code? Can we limit this?

File: Runtime/Components/WorldGen.cs [hunk 7]
```diff
@@ -237,16 +242,16 @@ private float getRatio(float pX, float pY)
...
-        StartCoroutine(BuildMeshSlow(Tx,Ty,m,RunInstant));
+		TilesBuilding.Add(i);
+		StartCoroutine(i);
 		//Should add the coroutine to a list, when the list is empty call the finished event
     }
 
-    private IEnumerator BuildMeshSlow(int Tx, int Ty, GameObject HoldsM,bool RunInstant = false)
+    private IEnumerator BuildMeshSlow(int Tx, int Ty, GameObject HoldsM, bool RunInstant = false)
     {
 		if(HoldsM){ClearChildren(HoldsM);}
```
Comment (Oggal) [inline Runtime/Components/WorldGen.cs:254]: This line goes against style

File: Runtime/Components/WorldGen.cs [hunk 8]
```diff
@@ -270,14 +276,15 @@ private IEnumerator BuildMeshSlow(int Tx, int Ty, GameObject HoldsM,bool RunInst
...
+                float Xi = (x * VertDis) - (TileSize / 2);
+                float Yi = (y * VertDis) - (TileSize / 2);
                 verts[id] = new Vector3(
                     Xi,
                     GetHeight(Xi + (lx), Yi + (ly), getRatio(Xi + lx, Yi + ly),MaxLOD),
                     Yi);
 #region UV map
-                uv[id] = CalcUV(x,y);
+                uv[id] = gridUV(x,y);
+				normals[id] = GetNormal(Xi+lx, Yi+ly);
```
Comment (Oggal) [inline Runtime/Components/WorldGen.cs]: Style Correction Needed.

File: Runtime/Components/WorldGen.cs [hunk 9]
```diff
@@ -290,8 +297,9 @@ private IEnumerator BuildMeshSlow(int Tx, int Ty, GameObject HoldsM,bool RunInst
 		m.vertices = verts;
         m.uv = uv;
         m.triangles = tri;
+		m.normals = normals;
        // yield return null;
-        m.RecalculateNormals();
+        //m.RecalculateNormals();
```
Comment (Oggal) [inline Runtime/Components/WorldGen.cs:300]: Should remove if not needed.

File: Runtime/Components/WorldGen.cs [hunk 10]
```diff
@@ -332,26 +340,39 @@ private IEnumerator BuildMeshSlow(int Tx, int Ty, GameObject HoldsM,bool RunInst
 		m.uv = uv;
 		m.triangles = tri;
 		// yield return null;
-		m.RecalculateNormals();
+		//m.RecalculateNormals();
```
Comment (Oggal) [inline Runtime/Components/WorldGen.cs]: this needs to be undone

File: Runtime/Components/WorldGen.cs [hunk 11]
```diff
@@ -422,7 +448,7 @@ private GameObject BuildTile(int Tx, int Ty, GameObject tile)
 		tile.transform.localScale = Vector3.one;
         tile.transform.position = Vector3.Scale(new Vector3(Tx * TileSize, 0, Ty * TileSize), transform.localScale);
 	
-		if (buildDecor)
+		if (buildDecor || SpawnPOIs)
 		{
 			BuildDecor(Tx,Ty,tile);
```
Comment (Oggal) [inline Runtime/Components/WorldGen.cs:451]: Separate decor and POI. PLEASE!!

File: Runtime/Components/WorldGen.cs [hunk 12]
```diff
@@ -477,18 +503,22 @@ public void BuildDecor(int Tx, int Ty, GameObject tile)
...
+		if (buildDecor)
+			Decor.AddRange(GetDecorinRect(TileArea));
+		if (GetComponent<POI_Gen>())
+			Decor.AddRange(GetComponent<POI_Gen>().GetPOIinRect(TileArea));
+		
+	
+			foreach (GameObject deco in Decor)
+			{
+				deco.transform.SetParent(tile.transform, true);
+			}
```
Comment (Oggal) [inline Runtime/Components/WorldGen.cs]: Indention is incorrect

**Review by Oggal — COMMENTED** [review]
Changes Needed. See other comments

File: Runtime/Components/WorldGen.cs (hunk 3 above)
Comment (Oggal) [reply Runtime/Components/WorldGen.cs:153]: This will be resolved next sprint

Comment (Oggal) [reply Runtime/Components/WorldGen.cs:153]: Make sure to get that added to the TODO list

**Review by Oggal — COMMENTED** [review]
Can't approve my own request, but I think I'm happy with the changes thus far.

The change to normals slows down the generation quite a bit but further improvements there are a task for a separate branch...

**Comment by 1stMan000** [comment]
Overall great work with the new updates. Though I strongly suggest dividing the functions. Functions should have a 10 lines max if possible with handling only one type of specific work

**Review by 1stMan000 — APPROVED** [review]
Overall great work with the new updates. Though I strongly suggest dividing the functions. Functions should have a 10 lines max if possible with handling only one type of specific work

**Pull request merged**